recipes/lora_finetune_distributed.py

```diff
@@ -307,16 +303,40 @@ def _setup_model(
                 base_model_state_dict_keys=base_model_state_dict.keys(),
             )
 
-            # Load both the base model weights and (if available) the adapter weights. Both
-            # of this should happen only on Rank 0
-            model.load_state_dict(base_model_state_dict, strict=False)
-            if lora_weights_state_dict:
-                model.load_state_dict(lora_weights_state_dict, strict=False)
+        self.adapter_params = get_adapter_params(model)
+        set_trainable_params(model, self.adapter_params)
 
-        else:
-            # For non-zero ranks, load the model on meta device
-            with utils.set_default_dtype(self._dtype), torch.device("meta"):
-                model = config.instantiate(cfg_model)
+        if enable_activation_checkpointing:
+            utils.set_activation_checkpointing(
+                model, auto_wrap_policy={modules.TransformerDecoderLayer}
+            )
+
+        for m in reversed(list(model.modules())):
+            if (
+                isinstance(m, nn.Linear)
+                and m.weight.requires_grad
+                or isinstance(m, modules.TransformerDecoderLayer)
+            ):
+                fully_shard(m)
+        fully_shard(model)
+
+        if lora_weights_state_dict:
+            utils.load_from_full_model_state_dict(
+                model, lora_weights_state_dict, self._device, self._is_rank_zero
+            )
+
+        with utils.set_default_dtype(self._dtype), self._device:
+            for m in model.modules():
+                if isinstance(m, LoRALinear) and not lora_weights_state_dict:
+                    m.lora_a.to_empty(device=self._device)
+                    m.lora_b.to_empty(device=self._device)
+                    m.initialize_parameters()
+                if isinstance(m, modules.RotaryPositionalEmbeddings):
```

Comment:
Just to clarify, we special handle RoPE because the buffer is not being loaded from a state dict, right?

Reply:
that's correct